flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/split/IcebergSourceSplit.java

```diff
@@ -147,7 +147,7 @@ byte[] serializeV2() throws IOException {
 
       for (FileScanTask fileScanTask : fileScanTasks) {
         String taskJson = FileScanTaskParser.toJson(fileScanTask);
-        out.writeUTF(taskJson);
+        writeBytes(out, taskJson);
```

Comment:
The reason why I don't use out.writeBytes here is because it is buggy. The position with buffer is already incremented by 1 [in here](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/core/memory/DataOutputSerializer.java#L161) ([Actual place](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/core/memory/DataOutputSerializer.java#L140) where the increment happens) and it is incremented by the whole size once again [in here](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/core/memory/DataOutputSerializer.java#L163) which makes it problematic.

So I had to take some inspiration from that code and introduce a custom version of it in the class further down.